src/cargo/core/package.rs

```diff
@@ -712,25 +704,31 @@ impl<'a, 'cfg> Downloads<'a, 'cfg> {
         dl.total.set(total);
         let now = Instant::now();
         if cur != dl.current.get() {
+            let delta = cur - dl.current.get();
+            let threshold = self.next_speed_check_bytes_threshold.get();
+
             dl.current.set(cur);
-            dl.updated_at.set(now);
+            self.updated_at.set(now);
 
-            if dl.current.get() >= dl.next_speed_check_bytes_threshold.get() {
-                dl.next_speed_check.set(now + dl.timeout.dur);
-                dl.next_speed_check_bytes_threshold.set(
-                    dl.current.get() + dl.timeout.low_speed_limit as u64,
+            if delta >= threshold {
+                self.next_speed_check.set(now + self.timeout.dur);
+                self.next_speed_check_bytes_threshold.set(
+                    self.timeout.low_speed_limit as u64,
                 );
+            } else {
+                self.next_speed_check_bytes_threshold.set(threshold - delta);
             }
         }
         if !self.tick(WhyTick::DownloadUpdate).is_ok() {
             return false
         }
 
         // If we've spent too long not actually receiving any data we time out.
-        if now - dl.updated_at.get() > dl.timeout.dur {
+        if now - self.updated_at.get() > self.timeout.dur {
```

Comment:
There is a bug here, but I'm not 100% sure what's wrong.  I'm doing some stress tests (5sec timeout), and I am getting this occasionally:
```
thread 'main' panicked at 'second instant is later than self', libcore/option.rs:1008:5
```
With some debugging, `self.updated_at` is just under 5 seconds greater than `now`.  I think what's happening is that `updated_at` is getting set to `now + timeout`, and then the `progress` function is getting called again with `cur` not being changed.  I don't think I understand why `updated_at` is set with a time in the future (considering it is commented as the last time bytes were received).  I would assume that `updated_at` should always be <= now.

I'm taking a break for a bit, so I probably won't look at this more until later today.